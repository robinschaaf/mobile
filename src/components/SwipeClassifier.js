import React from 'react'
import {
  StyleSheet,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import { connect } from 'react-redux'
import ClassificationPanel from './ClassificationPanel'
import Question from './Question'
import Tutorial from './Tutorial'
import Swipeable from './Swipeable'
import SwipeSubject from './SwipeSubject'
import SwipeTabs from './SwipeTabs'
import OverlaySpinner from './OverlaySpinner'
import NavBar from './NavBar'
import FullScreenImage from './FullScreenImage'
import { setState } from '../actions/index'
import { startNewClassification, setTutorialCompleted } from '../actions/classifier'
import { isEmpty } from 'ramda'

const mapStateToProps = (state, ownProps) => ({
  isFetching: state.classifier.isFetching,
  workflow: state.classifier.workflow[ownProps.workflowID] || {},
  project: state.classifier.project[ownProps.workflowID] || {},
  guide: state.classifier.guide[ownProps.workflowID] || {},
  tutorial: state.classifier.tutorial[ownProps.workflowID] || {},
  needsTutorial: state.classifier.needsTutorial[ownProps.workflowID] || false,
  subject: state.classifier.subject[ownProps.workflowID] || {},
  nextSubject: state.classifier.nextSubject[ownProps.workflowID] || {},
  subjectSizes: state.classifier.subjectSizes[ownProps.workflowID] || {},
  seenThisSession: state.classifier.seenThisSession[ownProps.workflowID] || [],
})

const mapDispatchToProps = (dispatch) => ({
  startNewClassification(workflowID) {
    dispatch(startNewClassification(workflowID))
  },
  setTutorialCompleted() {
    dispatch(setTutorialCompleted())
  },
  setIsFetching(isFetching) {
    dispatch(setState('classifier.isFetching', isFetching))
  },
})

export class SwipeClassifier extends React.Component {
  constructor(props) {
    super(props)
    this.setQuestionVisibility = this.setQuestionVisibility.bind(this)
    this.state = {
      isQuestionVisible: true,
      showFullSize: false,
    }
    this.props.setIsFetching(true)
    this.props.startNewClassification(this.props.workflowID)
  }

  setQuestionVisibility(isVisible) {
    this.setState({isQuestionVisible: isVisible})
  }

  finishTutorial() {
    if (this.props.needsTutorial) {
      this.props.setTutorialCompleted()
    } else {
      this.setQuestionVisibility(true)
    }
  }

  static renderNavigationBar() {
    return <NavBar title={'Classify'} showBack={true} />;
  }

  render() {
    const renderClassifierOrTutorial = () => {
      const key = this.props.workflow.first_task //always just one task
      const task = this.props.workflow.tasks[key]
      const allowPanAndZoom = this.props.workflow.configuration.pan_and_zoom

      const backSubject =
        <SwipeSubject
          inFront={false}
          subject={this.props.nextSubject}
          subjectSizes={this.props.subjectSizes}
          seenThisSession={this.props.seenThisSession}
          setImageSizes={() => {}}
        />

      const classification =
        <View>
          <Question
            question={task.question}
            workflowID={this.props.workflowID}
            taskHelp={task.help}
          />
          <View style={[styles.subjectContainer, { width: this.props.subjectSizes.resizedWidth, height: this.props.subjectSizes.resizedHeight }]}>
            { isEmpty(this.props.nextSubject) ? null : backSubject }
          </View>
        </View>

      const swipeableSubject =
        <Swipeable
          key={this.props.subject.id}
          workflowID={this.props.workflowID}
          showFullSize={() => this.setState({showFullSize: true})}
        />

      const tutorial =
        <Tutorial
          projectName={this.props.project.display_name}
          isInitialTutorial={this.props.needsTutorial}
          tutorial={this.props.tutorial}
          finishTutorial={() => this.finishTutorial()}
        />

      //swipeable subject needs to be outside the Classification Panel because of Android
      //the swipeable area is cut off to the panel otherwise
      const classificationPanel =
        <View style={{...StyleSheet.absoluteFillObject}}>
          <ClassificationPanel
            isFetching={ this.props.isFetching }
            hasTutorial = { !isEmpty(this.props.tutorial) }
            isQuestionVisible = {this.state.isQuestionVisible }
            setQuestionVisibility = { this.setQuestionVisibility }>
            { this.state.isQuestionVisible ? classification : tutorial }
          </ClassificationPanel>
          { this.state.isQuestionVisible ? swipeableSubject : null }
          { this.state.isQuestionVisible ? <SwipeTabs guide={this.props.guide} /> : null }
          <FullScreenImage
            source={{uri: this.props.subject.display.src}}
            isVisible={this.state.showFullSize}
            allowPanAndZoom={allowPanAndZoom}
            handlePress={() => this.setState({ showFullSize: false })} />
        </View>

      //needsTutorial is for the first time a guest or user visits this project
      return (
        this.props.needsTutorial ? tutorial : classificationPanel
      )
    }

    return (
      this.props.isFetching || isEmpty(this.props.workflow) ? <OverlaySpinner overrideVisibility={this.props.isFetching} /> : renderClassifierOrTutorial()
    )
  }
}

const styles = EStyleSheet.create({
  subjectContainer: {
    alignSelf: 'center',
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: 20,
  },
})

SwipeClassifier.propTypes = {
  isFetching: React.PropTypes.bool,
  workflowID: React.PropTypes.string,
  workflow: React.PropTypes.shape({
    first_task: React.PropTypes.string,
    tasks: React.PropTypes.object,
    configuration: React.PropTypes.object,
  }),
  subject: React.PropTypes.shape({
    id: React.PropTypes.string,
    display: React.PropTypes.shape({
      src: React.PropTypes.string
    })
  }),
  nextSubject: React.PropTypes.shape({
    id: React.PropTypes.string
  }),
  subjectSizes: React.PropTypes.object,
  seenThisSession: React.PropTypes.array,
  startNewClassification: React.PropTypes.func,
  setIsFetching: React.PropTypes.func,
  project: React.PropTypes.shape({
    display_name: React.PropTypes.string,
  }),
  tutorial: React.PropTypes.object,
  needsTutorial: React.PropTypes.bool,
  guide: React.PropTypes.object,
  setTutorialCompleted: React.PropTypes.func,
}

export default connect(mapStateToProps, mapDispatchToProps)(SwipeClassifier)
