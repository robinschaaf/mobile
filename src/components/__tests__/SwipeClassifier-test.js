import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

jest.mock('WebView', () => 'WebView')
jest.mock('../OverlaySpinner', () => 'OverlaySpinner')
jest.mock('../Question', () => 'Question')
jest.mock('../Tutorial', () => 'Tutorial')
jest.mock('../Swipeable', () => 'Swipeable')
jest.mock('../SwipeSubject', () => 'SwipeSubject')
jest.mock('../SwipeTabs', () => 'SwipeTabs')

import { SwipeClassifier } from '../SwipeClassifier'

const workflow = {
  first_task: 'T0',
  tasks: {
    T0: {
      question: 'What was that?',
      answers: [
        { label: 'Yes' },
        { label: 'No' },
      ]
    }
  }
}

const project = {
  display_name: 'Awesome project'
}

const subjectSizes={resizedWidth: 100, resizedHeight: 100}

const seenThisSession=[]

it('renders correctly', () => {
  const tree = renderer.create(
    <SwipeClassifier
      isFetching={false}
      setIsFetching={jest.fn}
      startNewClassification={jest.fn}
      saveThenStartNewClassification={jest.fn}
      saveAnnotation={jest.fn}
      project={project}
      workflow={workflow}
      workflowID={'1'}
      subject={{id: '23432432'}}
      subjectSizes={subjectSizes}
      seenThisSession={seenThisSession} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders spinner if fetching', () => {
  const tree = renderer.create(
    <SwipeClassifier
      isFetching={true}
      setIsFetching={jest.fn}
      startNewClassification={jest.fn}
      project={project}
      workflow={workflow}
      workflowID={'1'}
      subject={{id: '23432432'}}
      subjectSizes={subjectSizes}
      seenThisSession={seenThisSession} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders tutorial if needed', () => {
  const tree = renderer.create(
    <SwipeClassifier
      isFetching={false}
      setIsFetching={jest.fn}
      startNewClassification={jest.fn}
      needsTutorial={true}
      project={project}
      workflow={workflow}
      workflowID={'1'}
      subject={{id: '23432432'}}
      subjectSizes={subjectSizes}
      seenThisSession={seenThisSession} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
