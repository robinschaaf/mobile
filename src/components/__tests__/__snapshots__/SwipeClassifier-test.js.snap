exports[`test renders correctly 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }>
  <View
    style={undefined}>
    <View
      style={undefined}>
      <View
        style={undefined}>
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Array [
              undefined,
            ]
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                undefined,
              ]
            }>
            QUESTION
          </Text>
        </TouchableOpacity>
        <TouchableOpacity
          activeOpacity={0.2}
          onPress={[Function]}
          style={
            Array [
              undefined,
              undefined,
            ]
          }>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Array [
                undefined,
              ]
            }>
            TUTORIAL
          </Text>
        </TouchableOpacity>
      </View>
      <View>
        <Question
          question="What was that?"
          taskHelp={undefined}
          workflowID="1" />
        <View
          style={
            Array [
              undefined,
              Object {
                "height": 100,
                "width": 100,
              },
            ]
          }>
          <SwipeSubject
            inFront={false}
            seenThisSession={Array []}
            setImageSizes={[Function]}
            subject={undefined}
            subjectSizes={
              Object {
                "resizedHeight": 100,
                "resizedWidth": 100,
              }
            } />
        </View>
      </View>
    </View>
  </View>
  <Swipeable
    showFullSize={[Function]}
    workflowID="1" />
  <SwipeTabs
    guide={undefined} />
  <Modal
    animationType="fade"
    onRequestClose={[Function]}
    transparent={true}
    visible={false}>
    <View
      style={undefined}>
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "justifyContent": "center",
              "overflow": "hidden",
            },
            Object {
              "height": 667,
              "width": 375,
            },
          ]
        }>
        <View
          style={
            Object {
              "transform": Array [
                Object {
                  "scale": 1,
                },
                Object {
                  "translateX": 0,
                },
                Object {
                  "translateY": 0,
                },
              ],
            }
          }>
          <View
            onLayout={[Function]}
            style={
              Object {
                "height": 0,
                "width": 0,
              }
            }>
            <Image
              source={
                Object {
                  "uri": "blah.jpg",
                }
              }
              style={
                Object {
                  "height": 0,
                  "width": 0,
                }
              } />
          </View>
        </View>
      </View>
      <View
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontFamily": "FontAwesome",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }>
          
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              undefined,
              Array [
                undefined,
              ],
            ]
          }>
          You can zoom into this image
        </Text>
      </View>
      <TouchableOpacity
        activeOpacity={0.5}
        onPress={[Function]}
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": undefined,
                "fontFamily": "FontAwesome",
                "fontSize": 12,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }>
          
        </Text>
      </TouchableOpacity>
    </View>
  </Modal>
</View>
`;

exports[`test renders spinner if fetching 1`] = `
<OverlaySpinner
  overrideVisibility={true} />
`;

exports[`test renders tutorial if needed 1`] = `
<Tutorial
  finishTutorial={[Function]}
  isInitialTutorial={true}
  projectName="Awesome project"
  tutorial={undefined} />
`;
